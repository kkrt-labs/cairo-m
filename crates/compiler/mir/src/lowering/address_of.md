# AddressOf Operation Guidelines

## Overview

The AddressOf instruction is reserved for operations that require explicit
memory addresses, particularly for arrays and other memory-bound types. This
document outlines when and how AddressOf should be used.

## When to Use AddressOf

### Arrays (Future Implementation)

- Array element address calculation: `&arr[i]`
- Array slicing operations
- Passing arrays to functions that expect pointers

### Current Restrictions

- Tuples and structs should NOT use AddressOf for normal operations
- These types use value-based operations (MakeTuple, ExtractTuple, etc.)

## Implementation Pattern

```rust
// For future array implementation
match base_type {
    MirType::Array { .. } => {
        // Use AddressOf for array element references
        let elem_addr = self.calculate_element_address(base, index);
        self.instr().address_of(dest, elem_addr)
    }
    MirType::Tuple(_) | MirType::Struct { .. } => {
        // Use value-based operations, no AddressOf needed
        self.lower_value_aggregate_access(base, index)
    }
    _ => // handle other cases
}
```

## Memory Path vs Value Path

| Type   | Operations                                          | Path   |
| ------ | --------------------------------------------------- | ------ |
| Array  | framealloc, get_element_ptr, load, store, AddressOf | Memory |
| Tuple  | MakeTuple, ExtractTupleElement, InsertTuple         | Value  |
| Struct | MakeStruct, ExtractStructField, InsertField         | Value  |
