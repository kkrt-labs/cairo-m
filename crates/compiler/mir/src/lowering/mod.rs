//! # MIR Lowering Module
//!
//! This module contains the infrastructure for lowering the semantic AST to MIR.
//! It's organized into focused submodules that each handle a specific aspect of
//! the lowering process.

pub mod builder;
pub mod control_flow;
pub mod expr;
pub mod function;
pub mod stmt;
pub mod utils;

// Re-export the main entry point
pub use function::generate_mir;

// Re-export commonly used items
pub use builder::MirBuilder;
