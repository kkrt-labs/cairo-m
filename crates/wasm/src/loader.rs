//! WASM Module Loader
//!
//! This module provides functionality for loading and analyzing WASM modules,
//! with graceful handling of different womir API versions.

use std::fmt::{Debug, Display};
use std::fs;
use std::path::Path;

use thiserror::Error;

use womir::generic_ir::GenericIrSetting;
use womir::loader::{load_wasm, Program};

#[derive(Error, Debug)]
pub enum WasmLoadError {
    #[error("WASM file not found: {path}")]
    FileNotFound { path: String },
    #[error("Failed to read WASM file: {source}")]
    IoError { source: std::io::Error },
    #[error("Failed to parse WASM file: {message}")]
    ParseError { message: String },
}

/// A WASM module that lazily parses when first accessed
pub struct WasmModule {
    bytes: Vec<u8>,
}

impl WasmModule {
    /// Converts the WASM module into a WOMIR program.
    /// This is inefficient, as it will parse the WASM module everytime it is called.
    /// However we don't plan on using the WOMIR representation in the future.
    pub fn program(&self) -> Result<Program<'_, GenericIrSetting>, WasmLoadError> {
        load_wasm(GenericIrSetting, &self.bytes).map_err(|e| WasmLoadError::ParseError {
            message: e.to_string(),
        })
    }

    /// Loads a WASM module from a file.
    /// For now this just copies the bytes into the struct.
    pub fn from_file(file_path: &str) -> Result<Self, WasmLoadError> {
        let path = Path::new(file_path);
        if !path.exists() {
            return Err(WasmLoadError::FileNotFound {
                path: file_path.to_string(),
            });
        }

        let bytes = fs::read(path).map_err(|e| WasmLoadError::IoError { source: e })?;

        // Validate the bytes can be parsed (early error detection)
        load_wasm(GenericIrSetting, &bytes).map_err(|e| WasmLoadError::ParseError {
            message: e.to_string(),
        })?;

        Ok(Self { bytes })
    }
}

impl Display for WasmModule {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let program = match self.program() {
            Ok(prog) => prog,
            Err(e) => return write!(f, "Error parsing WASM: {}", e),
        };

        let mut output = String::new();

        for func in program.functions.iter() {
            // Get function name - check if exported, otherwise use default
            let func_name = program
                .c
                .exported_functions
                .get(&func.func_idx)
                .map(|name| name.to_string())
                .unwrap_or_else(|| format!("func_{}", func.func_idx));

            output.push_str(&format!(
                "Function: {} ({} directives)\n",
                func_name,
                func.directives.len()
            ));

            for (i, directive) in func.directives.iter().enumerate() {
                output.push_str(&format!("  {:3}: {}\n", i, directive));
            }
            output.push('\n'); // Empty line between functions
        }

        // Remove the trailing newline if there were functions
        if !output.is_empty() {
            output.pop(); // Remove the last newline
        }

        write!(f, "{}", output)
    }
}

impl Debug for WasmModule {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self)
    }
}
