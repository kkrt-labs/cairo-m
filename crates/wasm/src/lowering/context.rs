use super::DagToCasmError;
use cairo_m_compiler_codegen::{CasmBuilder, FunctionLayout};
use cairo_m_compiler_mir::{MirType, Value, ValueId};
use std::collections::HashMap;
use womir::loader::dag::ValueOrigin;

/// Context for converting a single function DAG to CASM
pub struct DagToCasmContext {
    pub(crate) casm_builder: CasmBuilder,
    /// Maps label IDs to their allocated stack slots for incoming values
    pub(crate) label_names: HashMap<u32, String>,
    pub(crate) label_slots: HashMap<u32, Vec<ValueId>>,
    pub(crate) loop_stack: Vec<ActiveLoop>,
    pub(crate) next_value_id: usize,
    pub(crate) value_types: HashMap<ValueId, MirType>,
    pub(crate) value_maps: Vec<HashMap<ValueOrigin, ValueId>>,
}

/// Information about an active loop during lowering
pub struct ActiveLoop {
    pub(crate) header_label: String,
    /// Stack slots allocated for loop-carried values
    pub(crate) header_slots: Vec<ValueId>,
}

impl DagToCasmContext {
    pub(crate) fn new(func_layout: FunctionLayout, label_counter: usize) -> Self {
        Self {
            casm_builder: CasmBuilder::new(func_layout, label_counter),
            label_names: HashMap::new(),
            label_slots: HashMap::new(),
            loop_stack: Vec::new(),
            next_value_id: 0,
            value_types: HashMap::new(),
            value_maps: vec![HashMap::new()], // Initialize with one scope for function-level values
        }
    }

    /// Generates a new unique value ID with type information
    pub fn new_typed_value_id(&mut self, mir_type: MirType) -> ValueId {
        self.next_value_id += 1;
        let id = ValueId::new(self.next_value_id);
        self.value_types.insert(id, mir_type);
        let _ = self.casm_builder.layout.allocate_local(id, 2);
        id
    }

    /// Generates a new unique value ID with type information, without allocating
    pub fn new_value_id_for_type(&mut self, mir_type: MirType) -> ValueId {
        self.next_value_id += 1;
        let id = ValueId::new(self.next_value_id);
        self.value_types.insert(id, mir_type);
        id
    }

    pub(crate) fn insert_value(&mut self, origin: ValueOrigin, value_id: ValueId) {
        if let Some(map) = self.value_maps.last_mut() {
            map.insert(origin, value_id);
        }
    }

    pub(crate) fn get_value(&self, origin: &ValueOrigin) -> Option<ValueId> {
        for map in self.value_maps.iter().rev() {
            if let Some(v) = map.get(origin) {
                return Some(*v);
            }
        }
        None
    }

    pub(crate) fn push_scope(&mut self) {
        self.value_maps.push(HashMap::new());
    }

    pub(crate) fn pop_scope(&mut self) {
        self.value_maps.pop();
    }

    /// Get MIR value for a WASM ValueOrigin
    pub(crate) fn get_input_value(
        &self,
        value_origin: &ValueOrigin,
    ) -> Result<Value, DagToCasmError> {
        self.get_value(value_origin).map_or_else(
            || {
                let available_count = self.value_maps.iter().map(|map| map.len()).sum::<usize>();
                Err(DagToCasmError::ValueMappingError {
                    function_name: self.casm_builder.layout.name.clone(),
                    node_idx: value_origin.node,
                    reason: format!(
                        "Value not found: node {}, output {}",
                        value_origin.node, value_origin.output_idx
                    ),
                    available_count,
                })
            },
            |value_id| Ok(Value::operand(value_id)),
        )
    }
}
