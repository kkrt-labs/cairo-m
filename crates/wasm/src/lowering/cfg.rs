use cairo_m_compiler_mir::{Terminator, Value, ValueId};
use womir::loader::blockless_dag::{BlocklessDag, BreakTarget, Node, Operation, TargetType};
use womir::loader::dag::ValueOrigin;

use super::{DagToMirContext, DagToMirError, wasm_type_to_mir_type};
use crate::loader::BlocklessDagModule;
use crate::lowering::context::ActiveLoop;

impl DagToMirContext {
    /// Pass 1: Preallocate all the blocks associated with DAG labels and loops
    /// Create phi nodes for label outputs that will be populated later
    pub(super) fn allocate_blocks_and_phi_nodes(
        &mut self,
        func: &BlocklessDag,
    ) -> Result<(), DagToMirError> {
        for (node_idx, node) in func.nodes.iter().enumerate() {
            if let Operation::Label { id } = &node.operation {
                let block_id = self.mir_function.add_basic_block();
                self.label_map.insert(*id, block_id);
                let mir_types = node
                    .output_types
                    .iter()
                    .map(|t| wasm_type_to_mir_type(t, "unknown", "label output"))
                    .collect::<Result<Vec<_>, _>>()?;
                let phi_value_ids: Vec<ValueId> = self.create_phi_nodes(block_id, &mir_types);

                // Map the label outputs to these phi nodes
                for (output_idx, &phi_id) in phi_value_ids.iter().enumerate() {
                    self.insert_value(
                        ValueOrigin {
                            node: node_idx,
                            output_idx: output_idx as u32,
                        },
                        phi_id,
                    );
                }
                self.label_phi_nodes.insert(*id, phi_value_ids);
            }
        }
        Ok(())
    }

    /// Pass 2: Generate MIR instructions from DAG nodes
    pub(super) fn generate_instructions_from_dag(
        &mut self,
        dag: &BlocklessDag,
        module: &BlocklessDagModule,
    ) -> Result<(), DagToMirError> {
        for (node_idx, node) in dag.nodes.iter().enumerate() {
            match &node.operation {
                Operation::Inputs => {}

                Operation::WASMOp(wasm_op) => {
                    // Convert WASM operation to MIR instruction
                    let mir_value = self.convert_wasm_op_to_mir(node_idx, wasm_op, node, module)?;
                    if let Some(mir_value) = mir_value {
                        self.insert_value(
                            ValueOrigin {
                                node: node_idx,
                                output_idx: 0,
                            },
                            mir_value,
                        );
                    }
                }

                Operation::Label { id } => {
                    let block_id = self.label_map.get(id).copied().ok_or_else(|| {
                        DagToMirError::InvalidControlFlow {
                            function_name: self.mir_function.name.clone(),
                            reason: format!("Label {:?} not found", id),
                            operation_context: "resolving label".to_string(),
                        }
                    })?;
                    self.set_current_block(block_id);
                }

                Operation::Br(target) => {
                    // This is either a jump or a return
                    let target_block = self.resolve_break_target(node_idx, node, target)?;
                    // Edge copies
                    self.record_edge_values(node, target)?;

                    let terminator = Terminator::jump(target_block);
                    self.get_current_block()?.set_terminator(terminator);
                    self.set_current_block(target_block);
                }

                Operation::BrIf(target) => {
                    // Conditional branch - in our DAG, the condition is the last input
                    let cond_idx = node.inputs.len().checked_sub(1).ok_or_else(|| {
                        DagToMirError::InvalidControlFlow {
                            function_name: self.mir_function.name.clone(),
                            reason: "BrIf without condition input".to_string(),
                            operation_context: "resolving BrIf condition".to_string(),
                        }
                    })?;
                    let condition_value = self.get_input_value(&node.inputs[cond_idx])?;
                    let target_block = self.resolve_break_target(node_idx, node, target)?;
                    let else_block = self.mir_function.add_basic_block();

                    // Edge copies on the taken edge
                    self.record_edge_values(node, target)?;

                    let terminator = Terminator::branch(condition_value, target_block, else_block);
                    self.get_current_block()?.set_terminator(terminator);
                    self.set_current_block(else_block);
                }

                Operation::BrIfZero(target) => {
                    // Inverted conditional branch - condition is the last input
                    let cond_idx = node.inputs.len().checked_sub(1).ok_or_else(|| {
                        DagToMirError::InvalidControlFlow {
                            function_name: self.mir_function.name.clone(),
                            reason: "BrIfZero without condition input".to_string(),
                            operation_context: "resolving BrIfZero condition".to_string(),
                        }
                    })?;
                    let condition_value = self.get_input_value(&node.inputs[cond_idx])?;
                    let else_target = self.resolve_break_target(node_idx, node, target)?;
                    let then_target = self.mir_function.add_basic_block();

                    // Edge copies on the taken edge
                    self.record_edge_values(node, target)?;

                    let terminator = Terminator::branch(condition_value, then_target, else_target);
                    self.get_current_block()?.set_terminator(terminator);
                    self.set_current_block(then_target);
                }

                Operation::BrTable { targets: _ } => {
                    todo!()
                }

                Operation::Loop {
                    sub_dag,
                    break_targets: _,
                } => {
                    // Build a normal loop (header/body/exit) from the sub-DAG
                    // Create header block and allocate phi nodes for loop-carried values
                    let header_block = self.mir_function.add_basic_block();

                    // Get loop input types from the sub-DAG's Inputs node
                    let sub_inputs_idx = 0;
                    let input_node = &sub_dag.nodes[sub_inputs_idx];
                    assert!(
                        matches!(input_node.operation, Operation::Inputs),
                        "Loop sub-DAG must start with Inputs node"
                    );

                    // Create phi nodes in the header for loop-carried values
                    let header_mir_types = input_node
                        .output_types
                        .iter()
                        .map(|t| wasm_type_to_mir_type(t, &self.mir_function.name, "loop input"))
                        .collect::<Result<Vec<_>, _>>()?;
                    let header_phi_nodes = self.create_phi_nodes(header_block, &header_mir_types);

                    // Record initial phi operands from loop entry
                    let current_block = self.current_block_id.unwrap();
                    for (input_idx, input) in node.inputs.iter().enumerate() {
                        if let Some(&phi_value_id) = header_phi_nodes.get(input_idx) {
                            let source_value_id = self.get_value(input).ok_or_else(|| {
                                let available_count =
                                    self.value_maps.iter().map(|map| map.len()).sum::<usize>();
                                DagToMirError::ValueMappingError {
                                    function_name: self.mir_function.name.clone(),
                                    node_idx: input.node,
                                    reason: format!(
                                        "Loop input {} (node {}, output {}) not found in value map",
                                        input_idx, input.node, input.output_idx
                                    ),
                                    available_count,
                                }
                            })?;
                            self.add_deferred_phi_operand(
                                header_block,
                                phi_value_id,
                                current_block,
                                Value::operand(source_value_id),
                            );
                        }
                    }

                    let terminator = Terminator::jump(header_block);
                    self.get_current_block()?.set_terminator(terminator);
                    self.set_current_block(header_block);

                    // Allocate a new block for the loop body
                    let body_block = self.mir_function.add_basic_block();
                    let terminator = Terminator::jump(body_block);
                    self.get_current_block()?.set_terminator(terminator);
                    self.set_current_block(body_block);

                    self.loop_stack.push(ActiveLoop {
                        header_block,
                        header_phi_nodes: header_phi_nodes.clone(),
                    });

                    // Enter a new value scope for the loop body to avoid ValueOrigin collisions
                    self.push_scope();

                    // Map the sub-DAG's Inputs node (node 0) to header phi nodes
                    for (output_idx, phi_value_id) in header_phi_nodes.iter().enumerate() {
                        self.insert_value(
                            ValueOrigin {
                                node: 0,
                                output_idx: output_idx as u32,
                            },
                            *phi_value_id,
                        );
                    }

                    // Pre-allocate labels inside the loop sub-DAG
                    self.allocate_blocks_and_phi_nodes(sub_dag)?;
                    // Lower the body
                    self.generate_instructions_from_dag(sub_dag, module)?;
                    // Exit the loop body's value scope
                    self.pop_scope();

                    // Pop loop and restore state
                    self.loop_stack.pop();
                }
            }
        }

        Ok(())
    }

    /// Helper: record phi operands for the taken edge of a branch based on BreakTarget kind
    pub(super) fn record_edge_values(
        &mut self,
        node: &Node,
        target: &BreakTarget,
    ) -> Result<(), DagToMirError> {
        match &target.kind {
            TargetType::Label(label_id) => self.record_branch_values_for_label(node, *label_id),
            TargetType::FunctionOrLoop => self.record_branch_values_for_loop(node, target.depth),
        }
    }

    /// Resolve a WASM break target to a MIR BasicBlockId
    pub(super) fn resolve_break_target(
        &mut self,
        node_idx: usize,
        node: &Node,
        target: &BreakTarget,
    ) -> Result<cairo_m_compiler_mir::BasicBlockId, DagToMirError> {
        match (&target.kind, target.depth) {
            (TargetType::Label(label_id), _depth) => {
                // Direct jump to a label at current scope
                self.label_map.get(label_id).map_or_else(
                    || {
                        Err(DagToMirError::InvalidControlFlow {
                            function_name: self.mir_function.name.clone(),
                            reason: format!("Label {} not found in label_map", label_id),
                            operation_context: "resolving break target".to_string(),
                        })
                    },
                    |block_id| Ok(*block_id),
                )
            }

            (TargetType::FunctionOrLoop, depth) => {
                // If inside a loop, this is a continue to the appropriate loop header.
                // depth == 0 => current loop, depth > 0 => outer loops
                let d = depth as usize;
                if !self.loop_stack.is_empty() && d < self.loop_stack.len() {
                    let idx = self.loop_stack.len() - 1 - d;
                    let loop_info = &self.loop_stack[idx];
                    Ok(loop_info.header_block)
                } else if d >= self.loop_stack.len() && !self.loop_stack.is_empty() {
                    return Err(DagToMirError::LoopDepthError {
                        function_name: self.mir_function.name.clone(),
                        node_idx,
                        requested_depth: depth,
                        available_depth: self.loop_stack.len(),
                    });
                } else {
                    // No active loop at this depth: treat as function return
                    let return_block = self.mir_function.add_basic_block();

                    let return_values: Vec<Value> = node
                        .inputs
                        .iter()
                        .map(|input| self.get_input_value(input))
                        .collect::<Result<_, _>>()?;

                    let terminator = Terminator::return_values(return_values);
                    self.block_mut(return_block)?.set_terminator(terminator);
                    Ok(return_block)
                }
            }
        }
    }

    /// Extract branch values from a node, excluding conditional inputs
    pub(super) fn get_branch_values(&self, node: &Node) -> Result<Vec<Value>, DagToMirError> {
        // Determine which inputs represent data (exclude condition for conditional branches)
        // For BrIf / BrIfZero, the last input is the condition; exclude it from data copies
        let data_inputs_start_index = 0;
        let data_inputs_end_index = match &node.operation {
            Operation::BrIf(_) | Operation::BrIfZero(_) => node.inputs.len().saturating_sub(1),
            _ => node.inputs.len(),
        };

        node.inputs[data_inputs_start_index..data_inputs_end_index]
            .iter()
            .map(|input| self.get_input_value(input))
            .collect()
    }

    /// Record branch values as phi operands for a label when branching to it
    pub(super) fn record_branch_values_for_label(
        &mut self,
        node: &Node,
        label_id: u32,
    ) -> Result<(), DagToMirError> {
        let branch_values = self.get_branch_values(node)?;
        let current_block =
            self.current_block_id
                .ok_or_else(|| DagToMirError::InvalidControlFlow {
                    function_name: self.mir_function.name.clone(),
                    reason: "No current block when recording phi operands".to_string(),
                    operation_context: "recording branch values for label".to_string(),
                })?;

        let phi_value_ids = self
            .label_phi_nodes
            .get(&label_id)
            .cloned()
            .ok_or_else(|| DagToMirError::InvalidControlFlow {
                function_name: self.mir_function.name.clone(),
                reason: format!("No phi nodes allocated for label {}", label_id),
                operation_context: "recording branch values for label".to_string(),
            })?;

        let target_block = self.label_map.get(&label_id).copied().ok_or_else(|| {
            DagToMirError::InvalidControlFlow {
                function_name: self.mir_function.name.clone(),
                reason: format!("Label {} not found in label_map", label_id),
                operation_context: "recording branch values for label".to_string(),
            }
        })?;

        // Record phi operands for each value
        let count = core::cmp::min(branch_values.len(), phi_value_ids.len());
        for i in 0..count {
            self.add_deferred_phi_operand(
                target_block,
                phi_value_ids[i],
                current_block,
                branch_values[i],
            );
        }

        Ok(())
    }

    /// Record branch values as phi operands for loop header when continuing to a loop
    pub(super) fn record_branch_values_for_loop(
        &mut self,
        node: &Node,
        depth: u32,
    ) -> Result<(), DagToMirError> {
        // Compute target loop based on depth
        let d = depth as usize;
        if self.loop_stack.is_empty() || d >= self.loop_stack.len() {
            return Ok(()); // Treat as return; phi operands handled by return elsewhere
        }
        let idx = self.loop_stack.len() - 1 - d;
        let loop_info = &self.loop_stack[idx];
        let header_phi_nodes = loop_info.header_phi_nodes.clone();
        let header_block = loop_info.header_block;

        let current_block =
            self.current_block_id
                .ok_or_else(|| DagToMirError::InvalidControlFlow {
                    function_name: self.mir_function.name.clone(),
                    reason: "No current block when recording phi operands".to_string(),
                    operation_context: "recording branch values for loop".to_string(),
                })?;

        let branch_values = self.get_branch_values(node)?;

        // Record phi operands for each loop-carried value
        let count = core::cmp::min(branch_values.len(), header_phi_nodes.len());
        for i in 0..count {
            self.add_deferred_phi_operand(
                header_block,
                header_phi_nodes[i],
                current_block,
                branch_values[i],
            );
        }

        Ok(())
    }
}
