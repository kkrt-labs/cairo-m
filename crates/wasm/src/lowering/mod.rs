//! Lowering from WOMIR BlockLess DAG to Cairo-M MIR.
mod cfg;
mod context;
mod ops;

use cairo_m_compiler_mir::{MirFunction, MirModule, MirType, PassManager};
use context::DagToMirContext;
use thiserror::Error;
use womir::loader::FunctionProcessingStage;
use womir::loader::dag::ValueOrigin;

use crate::loader::{BlocklessDagModule, WasmLoadError};

#[derive(Error, Debug)]
pub enum DagToMirError {
    #[error("Failed to load Wasm module: {0}")]
    WasmLoadError(#[from] WasmLoadError),
    #[error(
        "Unsupported WASM operation {op:?} in function '{function_name}' at node {node_idx}: {suggestion}"
    )]
    UnsupportedOperation {
        op: String,
        function_name: String,
        node_idx: usize,
        suggestion: String,
    },
    #[error("Invalid control flow in function '{function_name}': {reason}")]
    InvalidControlFlow {
        function_name: String,
        reason: String,
        operation_context: String,
    },
    #[error(
        "Value mapping error in function '{function_name}' at node {node_idx}: {reason} (available: {available_count} values)"
    )]
    ValueMappingError {
        function_name: String,
        node_idx: usize,
        reason: String,
        available_count: usize,
    },
    #[error("Unsupported WASM type {wasm_type:?} in function '{function_name}': {context}")]
    UnsupportedWasmType {
        wasm_type: wasmparser::ValType,
        function_name: String,
        context: String,
    },
    #[error(
        "Loop structure error in function '{function_name}' at node {node_idx}: depth {requested_depth} exceeds available {available_depth}"
    )]
    LoopDepthError {
        function_name: String,
        node_idx: usize,
        requested_depth: u32,
        available_depth: usize,
    },
}

/// Lower a whole WOMIR program to MIR
pub fn lower_program_to_mir(
    module: &BlocklessDagModule,
    mut pipeline: PassManager,
) -> Result<MirModule, DagToMirError> {
    let mut mir_module = MirModule::new();
    let program = &module.0;
    for (func_idx, _) in program.functions.iter().enumerate() {
        let mut mir_function = function_to_mir(module, func_idx)?;
        pipeline.run(&mut mir_function);
        mir_module.add_function(mir_function);
    }
    Ok(mir_module)
}

/// Convert WASM type to MIR type (limited support for now)
fn wasm_type_to_mir_type(
    wasm_type: &wasmparser::ValType,
    function_name: &str,
    context: &str,
) -> Result<MirType, DagToMirError> {
    match wasm_type {
        wasmparser::ValType::I32 => Ok(MirType::U32),
        _ => Err(DagToMirError::UnsupportedWasmType {
            wasm_type: *wasm_type,
            function_name: function_name.to_string(),
            context: context.to_string(),
        }),
    }
}

/// Convert a single WASM function to MIR using a two-pass algorithm
fn function_to_mir(
    module: &BlocklessDagModule,
    func_idx: usize,
) -> Result<MirFunction, DagToMirError> {
    let program = &module.0;
    let func_name = program
        .m
        .exported_functions
        .get(&(func_idx as u32))
        .map(|s| s.to_string())
        .unwrap_or_else(|| format!("func_{}", func_idx));

    let mut context = DagToMirContext::new(func_name.clone());

    // Get function type information for parameters and return types
    let func_type = program.m.get_func_type(func_idx as u32);

    // Handle param types with proper error handling
    let param_types: Vec<MirType> = func_type
        .ty
        .params()
        .iter()
        .map(|ty| wasm_type_to_mir_type(ty, &func_name, "function parameters"))
        .collect::<Result<Vec<MirType>, DagToMirError>>()?;

    // Handle return types with proper error handling
    let return_types: Vec<MirType> = func_type
        .ty
        .results()
        .iter()
        .map(|ty| wasm_type_to_mir_type(ty, &func_name, "function return types"))
        .collect::<Result<Vec<MirType>, DagToMirError>>()?;

    // Allocate parameters
    for (i, param_type) in param_types.iter().enumerate() {
        let param_id = context.mir_function.new_typed_value_id(param_type.clone());
        context.mir_function.parameters.push(param_id);
        context.insert_value(
            ValueOrigin {
                node: 0,
                output_idx: i as u32,
            },
            param_id,
        );
    }

    // Get the DAG for this function
    let dag = match program.functions.get(func_idx) {
        Some(FunctionProcessingStage::BlocklessDag(dag)) => dag,
        Some(_) => {
            return Err(DagToMirError::InvalidControlFlow {
                function_name: func_name,
                reason: "Function not at BlocklessDag stage".to_string(),
                operation_context: "lowering function".to_string(),
            });
        }
        None => {
            return Err(DagToMirError::ValueMappingError {
                function_name: func_name,
                node_idx: 0,
                reason: format!("Function {} not found", func_idx),
                available_count: program.functions.len(),
            });
        }
    };

    // Preallocate CFG structures and lower body
    context.allocate_blocks_and_phi_nodes(dag)?;
    context.generate_instructions_from_dag(dag, module)?;

    // Finalize all phi nodes with their operands
    context.finalize_phi_nodes()?;

    // Set entry block if we created any blocks
    if !context.mir_function.basic_blocks.is_empty() {
        context.mir_function.entry_block = 0.into();
    }

    // Populate parameter types
    for (i, &param_value_id) in context.mir_function.parameters.iter().enumerate() {
        if let Some(param_type) = param_types.get(i) {
            context
                .mir_function
                .value_types
                .insert(param_value_id, param_type.clone());
        }
    }

    // Define return values from the function signature (types/arity only).
    // The actual values returned are supplied by each Return terminator.
    context.mir_function.return_values = return_types
        .iter()
        .map(|ty| context.mir_function.new_typed_value_id(ty.clone()))
        .collect();

    Ok(context.mir_function)
}
