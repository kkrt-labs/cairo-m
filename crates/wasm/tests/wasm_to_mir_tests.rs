//! Main test runner for WASM to MIR conversion.

use insta::assert_snapshot;
use std::path::PathBuf;

// Use the loader from our src module
use cairo_m_compiler_mir::{PassManager, PrettyPrint};
use cairo_m_wasm::loader::BlocklessDagModule;
use cairo_m_wasm::lowering::lower_program_to_mir;

mod test_utils;
use test_utils::build_wasm;

/// A macro to define a WASM to MIR conversion test case
macro_rules! wasm_test {
    ($(#[$attr:meta])* $test_name:ident, $file_name:expr) => {
        #[test]
        $(#[$attr])*
        fn $test_name() {
            let wat_file_path = PathBuf::from(&format!("tests/test_cases/{}", $file_name));
            let wasm_file_path = wat_file_path.with_extension("wasm");

            if !wasm_file_path.exists() {
                build_wasm(&wat_file_path);
            }

            // Load the WASM module
            let wasm_file = std::fs::read(&wasm_file_path).unwrap();
            let module = BlocklessDagModule::from_bytes(&wasm_file).unwrap();
            // Lower to MIR without any optimizations
            let mir_module = lower_program_to_mir(&module, PassManager::no_opt_pipeline()).unwrap();

            // Create snapshot content
            let snapshot_content = {
                let module_output = mir_module.pretty_print(0);

                    format!(
                    "---\nsource: {}\nexpression: wasm_load_result\n---\nWASM File: {}\n============================================================\nSuccess: true\nFunctions loaded: {}\n============================================================\nModule Output:\n{}",
                    file!(),
                    $file_name,
                    mir_module.function_count(),
                    if module_output.is_empty() { "No functions found" } else { &module_output }
                )
            };

            // Snapshot the result
            assert_snapshot!(stringify!($test_name), snapshot_content);
        }
    };
}

// ====== Test Cases ======

// --- Basic WASM to MIR Conversion Tests ---
wasm_test!(convert_add_wasm, "add.wat");
wasm_test!(convert_fib_wasm, "fib.wat");
wasm_test!(convert_arithmetic_wasm, "arithmetic.wat");
wasm_test!(convert_simple_if_wasm, "simple_if.wat");
wasm_test!(convert_if_statement_wasm, "if_statement.wat");
wasm_test!(convert_func_call_wasm, "func_call.wat");
wasm_test!(convert_variables_wasm, "variables.wat");
wasm_test!(convert_bitwise_wasm, "bitwise.wat");
wasm_test!(convert_simple_loop_wasm, "simple_loop.wat");
wasm_test!(convert_nested_loop_wasm, "nested_loop.wat");
wasm_test!(convert_load_store_wasm, "load_store.wat");
wasm_test!(
    #[ignore]
    convert_sha256_wasm,
    "sha256.wat"
);
