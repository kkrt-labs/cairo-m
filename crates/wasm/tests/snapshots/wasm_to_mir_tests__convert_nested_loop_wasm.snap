---
source: crates/wasm/tests/wasm_to_mir_tests.rs
expression: snapshot_content
---
---
source: crates/wasm/tests/wasm_to_mir_tests.rs
expression: wasm_load_result
---
WASM File: nested_loop.wat
============================================================
Success: true
Functions loaded: 1
============================================================
Module Output:
module {
  // Function 0
  fn nested_loop {
    parameters: [0]
    entry: 0

    0:
      %2 = 0 (u32)
      %5 = %2 (u32)
      %4 = %2 (u32)
      %6 = %2 (u32)
      %3 = %0 (u32)
      jump 2

    1:
      jump 11

    2:
      jump 3

    3:
      %9 = %4 U32Less %3
      if %9 then jump 6 else jump 5

    4:
      %18 = 1 (u32)
      %19 = %4 U32Add %18
      %3 = %3 (u32)
      %4 = %19 (u32)
      %5 = %7 (u32)
      %6 = %8 (u32)
      jump 2

    5:
      %1 = %6 (u32)
      jump 1

    6:
      %10 = 0 (u32)
      %12 = %10 (u32)
      %13 = %6 (u32)
      %11 = %4 (u32)
      jump 7

    7:
      jump 8

    8:
      %14 = %12 U32Less %11
      if %14 then jump 10 else jump 9

    9:
      %8 = %13 (u32)
      %7 = %12 (u32)
      jump 4

    10:
      %15 = %13 U32Add %11
      %16 = 1 (u32)
      %17 = %12 U32Add %16
      %11 = %11 (u32)
      %12 = %17 (u32)
      %13 = %15 (u32)
      jump 7

    11:
      return %1

  }

}
